### PR TITLE
[BPF] ipv6 kube-proxy

### PR DESCRIPTION
**Pull request by tomastigera**
## Description

This PR introduces IPv6 version of kube-proxy and ipv6 versions of related bpf map types. Either v4 or v6 kube-proxy is instantiated depending on which ip version for the dataplane is selected.

refs https://github.com/projectcalico/calico/issues/4736

## Todos

- [ ] Tests
- [ ] Documentation
- [ ] Release note

## Release Note

```release-note
TBD
```

## Reminder for the reviewer

Make sure that this PR has the correct labels and milestone set.

Every PR needs one `docs-*` label.

- `docs-pr-required`: This change requires a change to the documentation that has not been completed yet.
- `docs-completed`: This change has all necessary documentation completed.
- `docs-not-required`: This change has no user-facing impact and requires no docs.

Every PR needs one `release-note-*` label.

- `release-note-required`: This PR has user-facing changes. Most PRs should have this label.
- `release-note-not-required`: This PR has no user-facing changes.

Other optional labels:

- `cherry-pick-candidate`: This PR should be cherry-picked to an earlier release. For bug fixes only.
- `needs-operator-pr`: This PR is related to install and requires a corresponding change to the operator.

### CODE REVIEW

File: felix/bpf/bpfmap/bpf_maps.go
Comment (fasaxc) [inline felix/bpf/bpfmap/bpf_maps.go:87]:
```suggestion
	getMap := func(v4, v6 func() maps.Map) maps.Map {
```

File: felix/bpf/bpfmap/bpf_maps.go
Comment (fasaxc) [inline felix/bpf/bpfmap/bpf_maps.go:94]:
```suggestion
	getMapWithExistsCheck := func(v4, v6 func() maps.MapWithExistsCheck) maps.MapWithExistsCheck {
```

File: felix/bpf/nat/maps.go
Comment (fasaxc) [inline felix/bpf/nat/maps.go:181]:
```suggestion
func (k FrontendKey) AffinityKeyCopy() FrontEndAffinityKeyInterface {
```

File: felix/bpf/proxy/syncer.go
Comment (fasaxc) [inline felix/bpf/proxy/syncer.go:206]: Think the API would be cleaner if you passed in a `KVFactory` interface with these functions as methods.  Then you'd just need two static instances of that interface and all the constructor calls would be a lot more concise.

File: felix/bpf/routes/map6.go
Comment (fasaxc) [inline felix/bpf/routes/map6.go:188]: Did this get used?

File: felix/bpf/proxy/kube-proxy.go
Comment (fasaxc) [inline felix/bpf/proxy/kube-proxy.go:251]: Is this putting both IPv4 and IPv6 in the same Trie? I'm not sure if the CIDR trie is robust to that; there are code paths that panic if they have to compare v4 to v6 IPs so it may only blow up if you've got IPv4 and IPv6s that start with the same prefix!

File: felix/bpf/proxy/syncer.go
Comment (tomastigera) [reply felix/bpf/proxy/syncer.go]: I only pass the family and the rest is internal to the syncer.

**Comment by marvin-tigera** [comment]
OK, I will merge the pull request when it's ready, leave the commits as is when I merge it, and leave the branch after I've merged it.

**Review by fasaxc — APPROVED** [review]

**Pull request merged**
